bootstore/src/trust_quorum/share_pkg.rs

```diff
@@ -0,0 +1,320 @@
+    // the protocol. This is here in case we decide otherwise, and becaus an
+    // epoch is used for disk encryption purposes.
+    pub epoch: u32,
+    pub threshold: u8,
+    pub share: Vec<u8>,
+
+    /// Digests of all 256 shares so that each sled can verify the integrity
+    /// of received shares.
+    ///
+    /// No need for expensive zeroizing
+    #[zeroize(skip)]
+    pub share_digests: Vec<Sha3_256Digest>,
+
+    /// Salt used for key derivation
+    /// Since we use the same key for all pkgs, we share the same salt
+    pub salt: [u8; 32],
+
+    /// Nonce used for encryption.
+    ///
+    /// We generate all pkgs during rack init and ensure uniqueness of all nonces
+    pub nonce: [u8; 12],
+
+    /// We include a distinct subset of unused shares for each sled in the
+    /// initial group. This allows the sled to hand out unique shares when a
+    /// new sled joins the cluster. We keep these encrypted so that a single
+    /// sled does not have enough unencrypted shares to unlock the rack
+    /// without participating in trust quorum.
+    ///
+    /// Each sled should keep local track of which shares it has already handed
+    /// out.
+    ///
+    /// Each sled derives the same encryption key, but uses the unique nonce
+    /// for encrypting its unique shares.
+    ///
+    /// No need for expensive zeroizing
+    #[zeroize(skip)]
+    pub encrypted_shares: Vec<u8>,
+}
+
+/// An analog to [`SharePkgV0`] for nodes that were added after rack
+/// initialization. There is no encrypted_shares or nonces because of this.
+#[derive(Clone, PartialEq, Serialize, Deserialize, Zeroize, ZeroizeOnDrop)]
+pub struct LearnedSharePkgV0 {
+    #[zeroize(skip)]
+    pub rack_uuid: Uuid,
+    // We aren't planning on doing any reconfigurations with this version of
+    // the protocol
+    pub epoch: u32,
+    pub threshold: u8,
+    pub share: Vec<u8>,
+
+    // Digests of all 256 shares so that each sled can verify the integrity
+    // of received shares.
+    //
+    // No need for expensive nonsense
+    #[zeroize(skip)]
+    pub share_digests: Vec<Sha3_256Digest>,
+}
+
+/// Create a package for each sled
+pub fn create_pkgs(
+    rack_uuid: Uuid,
+    n: usize,
+) -> Result<Secret<Vec<SharePkgV0>>, TrustQuorumError> {
+    let rack_secret = RackSecret::new();
+    let threshold = n / 2 + 1;
+    let epoch = 0;
+    // We always generate 255 shares to allow new sleds to come online
+    let total_shares = 255;
+    let shares_per_sled = 255 / n;
+    let shares = rack_secret.split(threshold, total_shares)?;
+    let share_digests = share_digests(&shares);
+    let mut salt = [0u8; 32];
+    OsRng.fill_bytes(&mut salt);
+    let cipher = derive_encryption_key(&rack_uuid, &rack_secret, &salt);
+    let mut pkgs = Vec::with_capacity(n);
+    for i in 0..n {
+        // Each pkg gets a distinct subset of shares
+        let mut iter =
+            shares.iter().skip(i * shares_per_sled).take(shares_per_sled);
```

Comment:
I'm thinking this loop could be marginally easier to follow by iterating over `shares.chunks(shares_per_sled)`:

```
for chunk in shares.chunks(shares_per_sled) {
        let iter = chunk.iter();
...
```

Reply:
Maybe, but I'd still need the index for the nonce, so I'd have to do something like:

`for (i, chunk) in shares.chunks(shares_per_sled).enumerate()`

I'm also not sure that I'd actually end up with the exact right number of chunks, since leftovers may be present in smaller chunks, which is why I went down the explicit count route.